### PR TITLE
smartPacket1

### PR DESCRIPTION
**Pull request by quantumbagel**
This is a heavily revised version of LightAPRS-1.0 for the Near Space Project.

---
This change is [Reviewable](https://reviewable.io/reviews/redercj/lightaprs-1.0/2)

### CODE REVIEW

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 1]
```diff
@@ -106,90 +171,201 @@ void setup() {
   APRS_useAlternateSymbolTable(alternateSymbolTable); 
   APRS_setSymbol(Symbol);
   //increase following value (for example to 500UL) if you experience packet loss/decode issues. 
-  APRS_setPreamble(350UL);  
+  APRS_setPreamble(500UL);
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino:174]: Did this show any improvement?

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 2]
```diff
@@ -106,90 +171,201 @@ void setup() {
...
             APRS_setPathSize(pathSize);
-        }
-      
-      //send status message every 60 minutes
-      if(gps.time.minute() == 30){               
-        sendStatus();       
+          }
+          
+          //send status message every 15 minutes
+          if((gps.time.minute() % 15) == 0) {
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino:241]: in a follow on commit, let's see if we can make the status message sending a bit smarter, especially in the cases where we don't have GPS lock.  For example, to let us know how many satellites we have.

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 3]
```diff
@@ -106,90 +171,201 @@ void setup() {
...
+          if((gps.time.minute() % 15) == 0) {               
+            sendStatus();       
+          } else {
+            sendLocation();
+          }
+
+          freeMem();
+          Serial.flush();
+        } // if time to tx
+        // sleepSeconds(BeaconWait-((millis-loop_start)/1000));
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino]: we can remove this line.

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 4]
```diff
@@ -28,7 +29,31 @@
 #define AprsPinInput  pinMode(12,INPUT);pinMode(13,INPUT);pinMode(14,INPUT);pinMode(15,INPUT)
 #define AprsPinOutput pinMode(12,OUTPUT);pinMode(13,OUTPUT);pinMode(14,OUTPUT);pinMode(15,OUTPUT)
 
-//#define DEVMODE // Development mode. Uncomment to enable for debugging.
+#define DEVMODE // Development mode. Uncomment to enable for debugging.
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino]: I think we want to commit code to master that does not enable the debugging.  we want to prepare it like it is going to be headed to space.

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 5]
```diff
@@ -106,90 +171,201 @@ void setup() {
...
+    }
 
-         sendLocation();
+    secsTillTx -= round((millis()-loop_start)/1000);
+  } else {
+    secsToCheckBatt--;
 
-      }
+    secsToCheckBatt -= (millis()-loop_start)/1000;
+    // sleepSeconds(BattWait-((millis-loop_start)/1000));
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino:262]: this needs to be reviewed.  we are changing secsToCheckBatt but never sleeping as a result.

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 6]
```diff
@@ -106,90 +171,201 @@ void setup() {
...
+  }
+  // Serial.println("Loop time in milliseconds->");
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino]: these two lines can be removed.

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 7]
```diff
@@ -106,90 +171,201 @@ void setup() {
...
   } else {
+    balloonDescendRepeat = 0;
+  }
+  if (balloonDescendRepeat >= numDescendChecks) {
+    balloonPopped = true;
+  }
+}
+void updateZone() {
+  
+  long a = gps.altitude.feet();
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino:283]: is there a performance cost to calling this function or is it just reading from memory?  if it costs in performance, we need to read it once into a variable, then use that variable.

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 8]
```diff
@@ -39,16 +64,56 @@ bool alternateSymbolTable = false ; //false = '/' , true = '\'
...
+
+struct txZones {
+  long secsForTx;
+  long altitude;
+  bool goingUp;
+};
+
+#define NUM_ZONES 8
+struct txZones zones[NUM_ZONES] = {
+  {1, 20000, true},
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino]: all of these have secsForTx set to 1.  let's set some real values.  

how about -> 60, 60, 30, 15, 30, 60, 30, 15 ?

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 9]
```diff
@@ -106,90 +171,201 @@ void setup() {
...
+          continue;
+          }
+      }
   }
   
+
 }
 
+void updateComment() {
+  comment[0] = ' ';
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino:313]: in the next commit, let's change this by creating a string for the sprintf like this:
<set local variables for sprintf>
sprintf(comment, "U/D: %s XHU: %s%%  XTEMP: %s MX%03d", <variables>)

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 10]
```diff
@@ -329,6 +507,7 @@ void updateTelemetry() {
   sprintf(telemetry_buff + 50, "%02d", gps.satellites.isValid() ? (int)gps.satellites.value() : 0);
   telemetry_buff[52] = 'S';
   telemetry_buff[53] = ' ';
+  telemetry_buff[54] =
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino]: this looks odd

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 11]
```diff
@@ -301,7 +479,7 @@ void updatePosition() {
 
 
 void updateTelemetry() {
- 
+
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino]: not needed

File: LightAPRS-hab/LightAPRS-hab.ino [hunk 12]
```diff
@@ -498,7 +677,6 @@ static void printFloat(float val, bool valid, int len, int prec)
   }
 #endif
 }
-
```
Comment (redercj) [inline LightAPRS-hab/LightAPRS-hab.ino:501]: not needed

**Review by redercj — CHANGES_REQUESTED** [review]

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 2 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino:241]: Ok

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 3 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino]: Done.

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 4 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino]: Done.

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 5 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino:262]: Can we talk about this?

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 6 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino]: Done

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 7 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino:283]: There is a veeeery small performance cost.

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 8 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino]: Done

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 9 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino:313]: Ok.

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 10 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino]: Resolved.

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 11 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino]: Ok

File: LightAPRS-hab/LightAPRS-hab.ino (hunk 12 above)
Comment (quantumbagel) [reply LightAPRS-hab/LightAPRS-hab.ino:501]: What is "not needed?" the last line?

**Review by redercj — COMMENTED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/redercj/lightaprs-1.0/2)* status: all files reviewed, 11 unresolved discussions (waiting on @moc-buhtig and @redercj)

**Comment by redercj** [comment]
Merging this with some items left to work on.

status-> how many satellites can be added

secsToCheckBatt needs to be made to work with new code

sprintf for entire comment vs setting each character.

**Pull request merged**